pkg/cache/snapshot.go

```diff
@@ -173,6 +187,13 @@ func (c *ClusterQueue) accumulateResources(cohort *Cohort) {
 			cohort.Usage[fName] = used
 		}
 		for res, val := range resUsages {
+			// Similar to cohort.RequestableResources, we accumulate the usage above the guaranteed resources,
+			// here we should remove the guaranteed quota as well for that part can not be borrowed.
+			val -= c.guaranteedQuota(fName, res)
```

Comment:
@kerthcet @B1F030 with this change, aren't the functions AddWorkload and RemoveWorkload in Snapshot now broken?

Reply:
The logic is a bit hard to follow, so I am not sure if preemption is broken. But, semantically, the calculations of Add/Remove Workload are now wrong, because they don't consider the guaranteed quotas.

Reply:
Let me take a look.

Reply:
> The logic is a bit hard to follow, so I am not sure if preemption is broken. But, semantically, the calculations of Add/Remove Workload are now wrong, because they don't consider the guaranteed quotas.

We consider the guaranteed quotas in function `workloadFits` before we decide whether to `AddWorkload` or not(based on `fits: bool`). In [pkg/scheduler/preemption/preemption.go](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/scheduler/preemption/preemption.go#L378):
when cohort is nil, means feature LendingLimit cannot be enabled, using the original logic;
when cohort is not nil, go into function `cq.UsedCohortQuota` and `cq.RequestableCohortQuota`, consider the featureGate and the guaranteed quotas.

Reply:
Yes, I understand that `workloadFits` considers the guaranteed quotas. However, after we call `RemoveWorkload`, the cohort usage is semantically wrong (because this function doesn't consider guaranteed quotas). So I'm not convinced that the next call to `workloadFits` would be correct.

Reply:
How about implementing a unit test for this situation?

Reply:
This is a bug I think and will impact the preemption. so first of all we should reproduce this bug with a new test. And yes, the logic here is convoluted ...

Reply:
Since we should compare directly in admitting for the sake of performance, I think we should refactor the `AddWorkload` and `RemoveWorkload` then.

Reply:
Thanks. I was planning to make a patch release next week. Do you think you can have a fix with unit tests by tomorrow?

Reply:
Sorry, this part of logic is kind of complex, and seems hard to reproduce it. Maybe this weekend or later, need more time.

Reply:
We've fixed that bug in #1770 ,PTAL @alculquicondor